### ISSUE TITLE
Process requirement for sufficient demand

### ISSUE THREAD

**Issue by RossTate** [WebAssembly/proposals#68]
The current situation we are currently dealing with is first and foremost due a shortcoming in process. We have a feature that is about to ship despite lacking any means to evaluate its design. WebAssembly has a far too significant backwards compatibility constraint to be shipping unevaluated features. To prevent future such predicaments, as a starting place I suggest the following revision to the process:

> As a requirement for advancing to Stage 3, *every* feature of the proposal should have approval from a substantial user base (where the appropriate scale/meaning of "substantial" and "user base" depends on the specifics of the proposal).

There are two primary reasons for this revision:
1. We should not be wasting everyone's time implementing features that are not currently needed.
2. We cannot evaluate whether the proposal's design for a feature effectively achieves its intended goal without a substantial user base to either demonstrate its utility or to provide feedback on how the proposal needs to be revised to meet their needs.

This is certainly standard process in both language design and software engineering, but as we have seen we have strayed from this standard process, and we are now seeing the uncertainty that creates: a feature everyone wants is being held hostage by a feature no one currently needs or even properly understands the impact of. Furthermore, it sounds like we are on track to repeat this mistake. Apparently typed function references are planned to advance to stage 2 or 3 despite the fact that the needs of at least one of the main users of typed function references has yet to be considered, e.g. typed object-oriented languages rely on typed function references in v-tables for efficient class-method dispatch and rely on variance of function types to support signature refinement in overridden methods, which is used to support a variety of patterns such as in fluent APIs and method chaining.

With this requirement in place, long ago the reference types proposal would have been pressured to separate external references from subtyping. Even if y'all didn't notice the disconnect between those two components of the proposal, I have been trying to call attention to these issues for years, but without this process in place I had no great way to appeal because I kept being told that *eventually* we would all see that this is necessary to solve all the problems (and years later we are still waiting for that eventuality, since none of the problems have been addressed yet). Furthermore, all these other proposals down the line would not have been made to depend on the reference types proposal. There is no need for subtyping to depend on the `anyref` vision or on external references. There is no need for type imports to depend on the `anyref` top type or even on subtyping (there are tons of ways to express constraints on subtyping, and reyling primarily on subtyping constraints is nowadays discouraged as they're both insufficiently expressive and overly complication). There is no need for exception handling to depend on external references, and you can imagine many applications of WebAssembly with exception handling that would not even want external references. There is not even a need for garbage collection to depend on `anyref` or external references; we demonstrated that even languages relying on a uniform representation would generally prefer to use a uniform representation that is specialized to the needs of the language over a useless type like `anyref`. So instead of `anyref` being a bottleneck of all these proposals, you get a bunch of independent composable proposals, one of which is external references.

But pulling back from the current situation, what do y'all think of the proposed revision that every feature of a Stage 3 proposal should have enough demand to warrant implementation and to enable proper evaluation?

**Comment by aheejin** [comment]
> There is no need for exception handling to depend on external references, and you can imagine many applications of WebAssembly with exception handling that would not even want external references.

Please. Your opinion is not the only one it matters. After endless back and forth in the E repo, this kind of assertion comes across as not only too assertive but even arrogant.

**Comment by tlively** [comment]
The current requirement closest to this proposal is the stage 4 entry requirement that "At least one toolchain implements the feature." We have limited data because not many proposals have moved to phase 4, but so far it has worked out. That being said, thinking more about the needs to toolchains and source languages earlier in the process can never be a bad thing.

**Comment by RossTate** [comment]
Oh shoot, I meant to clarify that I was not asserting the removal of `exnref`. I was just asserting that `exnref` is independent of `anyref`. Does that address your concern, @aheejin?

**Comment by aheejin** [comment]
`exnref` is an external reference, even if it is not a subtype of `anyref`.

**Comment by RossTate** [comment]
No. An external reference is something that is not interpreted by the wasm code at all (except through callbacks to the environment via imported functions). An `exnref` is an internal reference, with a meaning that is interpreted and utilized internally by the wasm program. Sorry for the lack of clarity.

**Comment by jgravelle-google** [comment]
Not having followed the subtyping story too closely, I don't have much of a dog in this race. With that in mind, I would like to highlight how hostile the tone of this issue comes across as. I also think you have a valid point, but it's buried under enough vitriol that it's hard to want to engage.

> The current situation we are currently dealing with is first and foremost due a shortcoming in process.

That is a bold claim. To me it reads "your process is wrong, and so obviously wrong that anyone can see it."

> We should not be wasting everyone's time implementing features that are not currently needed.

That is highly uncharitable. 1) "The work you are doing is a waste of time" and 2) it supposes that no justification exists for why a given proposal is important. No one is implementing proposals just for the hell of it here.

> This is certainly standard process in both language design and software engineering,

"This is obvious, but I must have been the first one to actually say something".

> a feature everyone wants is being held hostage by a feature no one currently needs or even properly understands the impact of.

Everyone and no one are the strongest and most sweeping of generalizations. It is difficult not to read that as hyperbolic and reactionary. Not only presumptuous to assume one can speak for everyone, not only dismissive to assume nobody has any ideas about the impact.

"held hostage", additionally, is a heavily emotion-laden phrase.

---

I could continue but you get the idea.

I would like to point out that web standards work is only possible through a great deal of social goodwill. Being kind to one another is actually a core component of the work we do. And it is real work, we have a tendency to be deeply passionate about the future of the web, and when our visions diverge the tensions can rise quite quickly. Which is why we need to be so cautious and careful about the manner in which we communicate.

**Comment by tlively** [comment]
> So far you have had a major user base for every feature, namely C/C++/LLVM implementors and developers. But now that's no longer case: C/C++/LLVM isn't utilizing the proposed subtyping. I think that is what's fundamentally different here

That makes sense. Another example is in the bulk operations proposal, where the memory operations were important for threading support in the toolchain, but the bulk table operations were not immediately useful. Now the proposal is being held up by a dependency required only for those bulk table operations.

I guess my takeaway here is that we need to be mindful of minimum useful functionality not just when we create new proposals but also when we split things off of existing proposals.

**Comment by jgravelle-google** [comment]
> hopefully I managed to address the hostility while still presenting the case.

Thanks. I think that reads much clearer now actually, well done.

> As we move forward with a proposal, we should evaluate whether our proposed design, with all its component parts, fulfills those motivations.

It might be proposal-specific what the evaluation criteria actually are. What we could do is require some explicit validation criteria, e.g. "implemented in LLVM and V8, reduces code size by >1%", as a minimum early-validation needed to advance to stage 3. Declaring such a criteria would be required to move to stage 2, and it helps make the development plan and goals of the proposal more concrete. For example, Interface Types would want at least 2 languages able to generate adapter functions and communicate a nontrivial data type with a differing ABI. These would be necessary-but-not-sufficient conditions and should improve our confidence.

**Comment by tlively** [comment]
I like that idea. It's very similar to an idea raised multiple times as a way to make the "web VMs" requirement more flexible and I saw that a similar mechanism was also recently adopted by WASI in their own phases process.

**Comment by jgravelle-google** [comment]
To clarify, I'm not saying we should loosen the web VM wording, or anything related to stage 4. Maybe we should, but we could independently have feature-specific requirements for earlier stages too.

**Comment by RossTate** [comment]
> Conversely, we should ask whether the components of the design have significant motivation, and if not, consider whether they should be a part of this proposal.

I think this emphasizes a point I meant to place more focus on. A proposal is a collection of features. Part of the process for advancing to a certain stage should include 1) an evaluation of the utility of each feature in isolation, and 2) a plan for evaluating the design of each feature in isolation. Of course, there will often be dependencies between features that should be accounted for, but part of the process should be evaluating how subcomponents of a feature might be carved out. (I mentioned in certain settings the idea that proposals might be subdivided into stages, each of which could be in different phases, and this process would help identify features that would be better pushed back to a later stage.)

(This comment is meant to be complementary to, not contradictory to, the above suggestions.)
